nntrainer/dataset/batch_queue.cpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   batch_queue.cpp
+ * @date   13 July 2021
+ * @brief  This file contains thread safe queue
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <batch_queue.h>
+#include <chrono>
+
+#include <nntrainer_error.h>
+
+using namespace std::literals::chrono_literals;
+
+namespace nntrainer {
+BatchQueue::BatchQueue(unsigned int queue_capacity_) :
```

Comment:
Please ensure to take queue_capacity as the number of elements (not number of batches) as queue capacity controls the amount of memory consumed by the data prefetching. This should not change when batchsize is increased or decreased.

Reply:
It will be amount of iterations batch queue will have, if we decide to change element, it will be number of elements. data buffer will set number `buffersize / batchsize`. batch queue is constructed inside a function that knows the immutable batch size, so.. this won't change in the point of batch queue.

Reply:
sure :)

Reply:
There will be the default value for queue_capacity and it can be defined by user. I wonder why do it needs to be <= buffersize? Of course, it will waste the memory but, I guess there is no need to constrain it. isn't it?

Reply:
Yes, there is no constraint from the queue. It is from the who is using the queue. But the queue capacity is not equal to buffer size I think it's worth noting it here :)